File: backend/app/schemas/roster_import.py
"""花名册导入相关 Pydantic v2 schema。"""

from __future__ import annotations

from datetime import date, datetime
from typing import Any, Optional

from pydantic import BaseModel, Field


# ============================================================
# 上传
# ============================================================

class RosterUploadResponse(BaseModel):
    """上传响应。"""
    batch_id: int
    batch_no: str
    file_name: str
    file_size: int
    file_sha256: str
    sheet_names: list[str] = []
    selected_sheet: Optional[str] = None
    header_row: int = 1
    total_rows: int = 0
    status: str
    message: str = ""
    suggestion: Optional[str] = None


# ============================================================
# 字段映射
# ============================================================

class ColumnMapping(BaseModel):
    """单列映射。"""
    source_header: str = Field(..., description="Excel 列名")
    source_header_normalized: str = Field(..., description="标准化列名")
    target_field_key: Optional[str] = Field(default=None, description="映射到的系统字段")
    is_mapped: bool = False
    is_required: bool = False
    is_unknown: bool = False
    save_as_alias: bool = False


class FieldDefinition(BaseModel):
    """系统字段定义。"""
    key: str
    label: str
    required: bool = False
    description: str = ""
    example_values: str = ""


class SaveMappingRequest(BaseModel):
    """保存映射请求。"""
    mappings: dict[str, Optional[str]] = Field(..., description="Excel列名 -> 系统字段")


class SaveMappingResponse(BaseModel):
    """保存映射响应。"""
    batch_id: int
    mapped_count: int
    unmapped_count: int
    unmapped_required: list[str] = []


# ============================================================
# 预览
# ============================================================

class RowPreviewItem(BaseModel):
    """单行预览数据。"""
    row_no: int
    employee_id: Optional[int] = None
    normalized_name: Optional[str] = None
    matched_employee_name: Optional[str] = None
    row_status: str
    diff_fields: list[dict] = []
    issues: list[dict] = []
    can_skip: bool = True
    actions: list[str] = []


class PreviewSummary(BaseModel):
    """导入预览汇总。"""
    new_count: int = 0
    update_count: int = 0
    unchanged_count: int = 0
    confirmation_count: int = 0
    error_count: int = 0
    skipped_count: int = 0
    blocker_count: int = 0
    warning_count: int = 0
    total_rows: int = 0


class PreviewResponse(BaseModel):
    """预览响应。"""
    batch_id: int
    mode: str
    summary: PreviewSummary
    rows: list[RowPreviewItem] = []
    field_definitions: list[FieldDefinition] = []
    current_mappings: dict[str, str] = {}


# ============================================================
# 问题处理
# ============================================================

class IssueItem(BaseModel):
    """导入问题项。"""
    id: int
    row_id: Optional[int] = None
    employee_id: Optional[int] = None
    field_key: Optional[str] = None
    issue_code: str
    severity: str
    title: str
    message: Optional[str] = None
    old_value: Optional[Any] = None
    new_value: Optional[Any] = None
    allowed_actions: list[str] = []
    resolution_status: str = "PENDING"
    row_no: Optional[int] = None
    employee_name: Optional[str] = None


class ResolveIssueRequest(BaseModel):
    """处理问题请求。"""
    action: str = Field(..., description="操作")
    value: Optional[Any] = Field(default=None, description="处理值")
    note: Optional[str] = Field(default=None, description="备注")


# ============================================================
# 提交
# ============================================================

class CommitConfirmation(BaseModel):
    """提交确认信息。"""
    will_create: int = 0
    will_update: int = 0
    will_create_changes: int = 0
    will_create_compensations: int = 0
    will_create_notes: int = 0
    will_create_confirmations: int = 0
    will_skip: int = 0


class CommitResponse(BaseModel):
    """提交响应。"""
    batch_id: int
    batch_no: str
    status: str
    message: str
    summary: Optional[PreviewSummary] = None


# ============================================================
# 导入历史
# ============================================================

class BatchListItem(BaseModel):
    """导入批次列表项。"""
    id: int
    batch_no: str
    mode: str
    file_name: str
    file_sha256: str
    batch_status: str
    total_rows: int
    new_count: int
    update_count: int
    skipped_count: int
    confirmation_count: int
    error_count: int
    warning_count: int
    created_at: datetime
    completed_at: Optional[datetime] = None
    failed_at: Optional[datetime] = None
    rolled_back_at: Optional[datetime] = None


class BatchListResponse(BaseModel):
    """导入批次列表响应。"""
    items: list[BatchListItem]
    total: int


# ============================================================
# 撤销
# ============================================================

class RollbackItem(BaseModel):
    """撤销预览项。"""
    employee_id: int
    employee_name: str
    action: str
    details: str
    reversible: bool
    conflict: bool = False
    conflict_reason: Optional[str] = None


class RollbackPreviewResponse(BaseModel):
    """撤销预览响应。"""
    batch_id: int
    batch_no: str
    can_rollback: bool
    items: list[RollbackItem] = []
    irreversible_operations: list[str] = []
    conflicts: list[str] = []


# ============================================================
# 列别名
# ============================================================

class ColumnAliasItem(BaseModel):
    """列别名。"""
    id: int
    source_header_normalized: str
    target_field_key: str
    enabled: bool


class ValueAliasItem(BaseModel):
    """值别名。"""
    id: int
    field_key: str
    source_value_normalized: str
    target_value: str


class CreateColumnAliasRequest(BaseModel):
    """创建列别名请求。"""
    source_header_normalized: str
    target_field_key: str


class CreateValueAliasRequest(BaseModel):
    """创建值别名请求。"""
    field_key: str
    source_value_normalized: str
    target_value: str

File: frontend/src/types/roster-import.ts
// 花名册导入相关的 TypeScript 类型定义
// 与后端 schemas/roster_import.py 对应

export interface RosterUploadResponse {
  batch_id: number
  batch_no: string
  file_name: string
  file_size: number
  file_sha256: string
  sheet_names: string[]
  selected_sheet: string | null
  header_row: number
  total_rows: number
  status: string
  message: string
  suggestion: string | null
}

export interface ColumnMapping {
  source_header: string
  source_header_normalized: string
  target_field_key: string | null
  is_mapped: boolean
  is_required: boolean
  is_unknown: boolean
  save_as_alias: boolean
}

export interface FieldDefinition {
  key: string
  label: string
  required: boolean
  description: string
  example_values: string
}

export interface SaveMappingRequest {
  mappings: Record<string, string | null>
}

export interface PreviewSummary {
  new_count: number
  update_count: number
  unchanged_count: number
  confirmation_count: number
  error_count: number
  skipped_count: number
  blocker_count: number
  warning_count: number
  total_rows: number
}

export interface DiffField {
  field_key?: string
  field_label?: string
  old_value?: unknown
  new_value?: unknown
  change_type?: string
}

export interface RowPreviewItem {
  row_no: number
  employee_id: number | null
  normalized_name: string | null
  matched_employee_name: string | null
  row_status: string
  diff_fields: DiffField[]
  issues: IssueItem[]
  can_skip: boolean
  actions: string[]
}

export interface IssueItem {
  id: number
  row_id: number | null
  employee_id: number | null
  field_key: string | null
  issue_code: string
  severity: string
  title: string
  message: string | null
  old_value: unknown
  new_value: unknown
  allowed_actions: string[]
  resolution_status: string
  resolution_action?: string | null
  resolution_value_json?: unknown
  resolution_note?: string | null
  resolved_at?: string | null
  row_no: number | null
  employee_name: string | null
  created_at?: string | null
  updated_at?: string | null
  // 推荐处理字段
  source_value?: string | null
  system_value?: string | null
  recommended_action?: string | null
  recommended_value?: unknown
  recommendation_text?: string | null
  auto_fixable?: boolean
  requires_business_update?: boolean
}

export interface PreviewResponse {
  batch_id: number
  mode: string
  summary: PreviewSummary
  rows: RowPreviewItem[]
  field_definitions: FieldDefinition[]
  current_mappings: Record<string, string>
}

export interface BatchListItem {
  id: number
  batch_no: string
  mode: string
  file_name: string
  file_sha256: string
  batch_status: string
  total_rows: number
  new_count: number
  update_count: number
  skipped_count: number
  confirmation_count: number
  error_count: number
  warning_count: number
  created_at: string
  completed_at: string | null
  failed_at: string | null
  rolled_back_at: string | null
}

export interface CommitConfirmation {
  will_create: number
  will_update: number
  will_create_changes: number
  will_create_compensations: number
  will_create_notes: number
  will_create_confirmations: number
  will_skip: number
}

export interface CommitResponse {
  batch_id: number
  batch_no: string
  status: string
  message: string
  summary: PreviewSummary | null
}

export interface ResolveIssueRequest {
  action: string
  value?: unknown
  note?: string
}

export interface RollbackItem {
  employee_id: number
  employee_name: string
  action: string
  details: string
  reversible: boolean
  conflict: boolean
  conflict_reason: string | null
}

export interface RollbackPreviewResponse {
  batch_id: number
  batch_no: string
  can_rollback: boolean
  items: RollbackItem[]
  irreversible_operations: string[]
  conflicts: string[]
}

export interface BatchResolveRequest {
  issue_ids: number[]
  action: 'APPLY_RECOMMENDATION' | 'IGNORE'
  note?: string
}

export interface BatchResolveResponse {
  total: number
  resolved: number
  skipped: number
  conflicts: Array<{
    issue_id: number
    issue_code: string
    reason: string
  }>
  failed_ids: number[]
}

export interface ColumnAliasItem {
  id: number
  source_header_normalized: string
  target_field_key: string
  enabled: boolean
}

export interface ValueAliasItem {
  id: number
  field_key: string
  source_value_normalized: string
  target_value: string
}
